src/number.rs

```diff
@@ -98,22 +98,20 @@ fn try_parse_19digits(s: &mut AsciiStr<'_>, x: &mut u64) {
 }
 
 #[inline]
-fn try_parse_8digits_le(s: &mut AsciiStr<'_>, x: &mut u64) {
+fn try_parse_8digits(s: &mut AsciiStr<'_>, x: &mut u64) {
     // may cause overflows, to be handled later
-    if cfg!(target_endian = "little") {
-        if let Some(v) = s.try_read_u64() {
-            if is_8digits_le(v) {
-                *x = x
-                    .wrapping_mul(1_0000_0000)
-                    .wrapping_add(parse_8digits_le(v));
-                s.step_by(8);
-                if let Some(v) = s.try_read_u64() {
-                    if is_8digits_le(v) {
-                        *x = x
-                            .wrapping_mul(1_0000_0000)
-                            .wrapping_add(parse_8digits_le(v));
-                        s.step_by(8);
-                    }
+    if let Some(v) = s.try_read_u64() {
+        if is_8digits(v) {
+            *x = x
+                .wrapping_mul(1_0000_0000)
+                .wrapping_add(parse_8digits_le(v));
```

Comment:
Hm, I'm just trying to wrap my head around to what  exactly gets passed to `parse_8digits_le()` on big-endian arch and why does it actually work if it does...

Reply:
I guess it does work indeed, just isn't immediately obvious. Maybe `_le()` suffix should be removed then?

Reply:
Yes it should, thats my mistake. Ill fix it and recommit.

Reply:
@aldanor 

> Hm, I'm just trying to wrap my head around to what exactly gets passed to `parse_8digits_le()` on big-endian arch and why does it actually work if it does...

This works because of the following:

1. The code all assumes the bytes are read in little-endian order. Say I have the string "12345678", in little-endian, I have now read this as `0x3837363534333231`, or essentially, the string is reversed. 
2. The rest of the code is all pure numerical operations.

This is all the relevant code, besides `try_readu64`. There is only mathematical operations, that is, no matter what endian the architecture is, you get the same result as long as the integer passed to each function is the same. `is_8digits(0x3837363534333231)` will always produce the same result, as will `parse_8digits(0x3837363534333231)`, since there is no endian-dependent code paths for either.

What matters, however, is that the bytes are read (or written) in little-endian order in all cases. So merely changing `read_u64` should fix the issue in all cases.

```rust
#[inline]
pub fn is_8digits(v: u64) -> bool {
    let a = v.wrapping_add(0x4646_4646_4646_4646);
    let b = v.wrapping_sub(0x3030_3030_3030_3030);
    (a | b) & 0x8080_8080_8080_8080 == 0
}

#[inline]
fn parse_8digits(mut v: u64) -> u64 {
    const MASK: u64 = 0x0000_00FF_0000_00FF;
    const MUL1: u64 = 0x000F_4240_0000_0064;
    const MUL2: u64 = 0x0000_2710_0000_0001;
    v -= 0x3030_3030_3030_3030;
    v = (v * 10) + (v >> 8); // will not overflow, fits in 63 bits
    let v1 = (v & MASK).wrapping_mul(MUL1);
    let v2 = ((v >> 16) & MASK).wrapping_mul(MUL2);
    ((v1.wrapping_add(v2) >> 32) as u32) as u64
}

#[inline]
fn try_parse_8digits(s: &mut AsciiStr<'_>, x: &mut u64) {
    // may cause overflows, to be handled later
    if let Some(v) = s.try_read_u64() {
        if is_8digits(v) {
            *x = x
                .wrapping_mul(1_0000_0000)
                .wrapping_add(parse_8digits(v));
            s.step_by(8);
            if let Some(v) = s.try_read_u64() {
                if is_8digits(v) {
                    *x = x
                        .wrapping_mul(1_0000_0000)
                        .wrapping_add(parse_8digits(v));
                    s.step_by(8);
                }
            }
        }
    }
}
```

The only other endian-dependent path previously was this:

```rust
let v = s.read_u64();
if !is_8digits(v) {
    break;
}
d.digits[d.num_digits..].write_u64(v - 0x3030_3030_3030_3030);
d.num_digits += 8;
s = s.advance(8);
```

Let's say we have `b"12345678"`, then `s.read_u64()` will produce `0x3837363534333231` since we read it in little-endian order on all architectures (using `u64::from_le`). Then, `write_u64` will write `0x0807060504030201` as `[1, 2, 3, 4, 5, 6, 7, 8]` on all architectures, since we use `u64::to_le` internally. That is, we get the same input and output in all cases.

Reply:
Here's a simple program that proves the latter is true:

```rust
use std::ptr;

#[inline]
fn read_u64(bytes: &[u8]) -> u64 {
    debug_assert!(bytes.len() >= 8);
    let src = bytes.as_ptr() as *const u64;
    u64::from_le(unsafe { ptr::read_unaligned(src) })
}

#[inline]
fn write_u64(bytes: &mut [u8], value: u64) {
    debug_assert!(bytes.len() >= 8);
    let dst = bytes.as_mut_ptr() as *mut u64;
    unsafe { ptr::write_unaligned(dst, u64::to_le(value)) };
}

pub fn main() {
    let input = b"12345678";
    let value = read_u64(input);
    assert_eq!(value, 0x3837363534333231);
    let mut output = [0u8; 8];
    write_u64(&mut output, value - 0x3030_3030_3030_3030);
    assert_eq!(output, [1, 2, 3, 4, 5, 6, 7, 8]);
}
```

The former example should be much easier to conceptualize. This runs perfectly fine on both little-endian and big-endian systems.

I hope this makes sense?

Reply:
This is pretty much what I reproduced in my head as well, thanks for making it crystal clear and putting it in writing!